man/bppbranchlik.1
.\" SPDX-FileCopyrightText: The Bio++ Development Group
.\"
.\" SPDX-License-Identifier: CECILL-2.1

.TH BPPBRANCHLIK 1 LOCAL

.SH NAME

bppbranchlik - Computation of per branch (and per site) log-likelihood
of a model (ie only on this branch) in another model background (ie on
all other branches).

.SH SYNOPSIS

.B bppbranchlik [options]

.SH AVAILABILITY

All UNIX flavors

.SH DESCRIPTION

Given a phylo-likelihood computed on a process or a set of processes
in a Partition, the model(s) of another process -- aka the alternative
process -- are applied on all branches the one after the other, and
(site-)log-likelihoods are computed given this change. 

.SH OPTIONS

You should refer to 'info bppsuite' or to the online manual of
bppsuite for a complete list of available options.

.TP 5

--noninteractive

generates output for redirection in a file.

.TP

param=file

reads a file for loading options

.SH AUTHOR

The Bio++ Development Team.


ChangeLog
19/03/18 -*- Version 2.4.0 -*-

19/02/18 Julien Dutheil
* Bug fixed in bppPopStats: dN and dS were inverted!

10/12/17 -*- Version 2.3.2 -*-

06/11/17 Julien Dutheil
* Added estimation of kappa to bppPopStats + bugs fixed

06/06/17 -*- Version 2.3.1 -*- 

10/05/17 -*- Version 2.3.0 -*- 

21/12/16 Julien Dutheil
* BppPhySamp migrated to its own package.

17/12/16 Julien Dutheil
* Improved bppPopStats, outgroup analyses, MK table and output to file.

06/10/16 Julien Dutheil
* bppSeqGen now takes Hudson's MS ARG files as input.
* new input.tree.scale option to adjust branch lengths.

22/05/16 Julien Dutheil
* bpppopstats now computes codon statistics per site (e.g. PiN, PiS)
* Added example for bpppopstats.

29/06/15 Julien Dutheil
* New program bppPopStats

28/01/15 Julien Dutheil
* PhySamp now outputs the correct tree!

25/11/14 Julien Dutheil
* bppalnscore output contiguous segments in mase selection.
* bppalnscore site selection works with codon alphabets.

24/11/14 Julien Dutheil
* bppSeqMan can now read alignments and perform site selection priori to any analysis.

28/09/14 -*- Version 2.2.0 -*- 

16/09/14 Julien Dutheil
* bppSeqGen supports generic characters as ancestral states in info file (bug #87).

12/08/14 Julien Dutheil
* bppSeqGen supports simulation under Markov-Modulated models.

06/06/14 Julien Dutheil
* bppPhySamp now output trees as well.

17/04/14 Julien Dutheil
* bppSeqMan: Translate uses global genetic_code argument instead of local 'code' argument.
* All programs use the --seed argument to set random seed for reproduceable results.

08/03/13 -*- Version 0.8.0 -*-

22/01/13 Julien Dutheil
* Compiles with -Wconversion

14/01/13 Mathieu Groussin
* Added COaLA model for proteins in bppML and bppSeqGen.

23/11/12 Laurent Guéguen
* New bppMixedLikelihood program

08/02/12 -*- Version 0.7.0 -*-

21/12/11 Julien Dutheil
* New bppAlnScore program
* Improved bppSeqMan: protein translation now works + more stop codons removal options.

17/06/11 Julien Dutheil
* Improved bppancestor, also allows to compute frequencies for leaves.

11/06/11 -*- Version 0.6.2 -*-
* Small bug fixed + new packages.

03/06/10 -*- Version 0.6.0 -*-
* Compatibility with Bio++ 2.0.0.

03/06/10 Julien Dutheil
* New method in bppPhySamp + new, faster, implementation.

25/03/10 -*- Version 0.5.0 -*-

22/08/09 Julien Dutheil
* Added CMake configuration files, with twick for building the manual (assumes
		texinfo is installed).

22/07/09 Julien Dutheil
* Added program bppTreeDraw.

30/06/09 Julien Dutheil
* BppAncestor: now output expected ancestral states frequencies.
* BppAncestor: new method to compute EB estimates of ancestral states frequencies.

10/06/09 -*- Version 0.4.0 -*-

22/05/09 Julien Dutheil
* Keyval syntax for sequence format description.
* Documentation and example update.
* SeqMan now uses the SequenceApplicationTools method.

15/05/09 Julien Dutheil
* Compatibility update with the libs, new keyval syntax for model
specification.
* Documentation up to date.

02/02/09 Julien Dutheil
* bppConsense now outputs the number of leaves of the input tree.
* New progress bar in bppConsense.

08/01/09 Julien Dutheil
* Added texinfo documentation.
* More bootstrap options in BppDist (the same as in BppML).

25/09/08 -*- Version 0.3.1 -*-

07/12/08 Bastien Boussau & Julien Dutheil
* Bug removed in non-homogeneous models with covarions.

09/10/08 Julien Dutheil
* New action 'invert' in bppSeqMan.

25/09/08 -*- Version 0.3.0 -*-

20/09/08 Julien Dutheil
* New program bppAncestor to reconstruct acestral states and rates.
* New +F models from PhylLib. 

02/07/08 Laurent Guegen
* Bug fixed in bppML: frequencies can be computed from the data with
non-homogeneous models.

19/06/08 Julien Dutheil
* Shell variables are now 'global' in option files.

24/04/08 -*- Version 0.2.3 -*-
* This version requires the CVS version of Bio++ to be built.

03/03/08 Julien Dutheil
* Bug fixed in keep_complete method of BppSeqMan.

27/03/08 Julien Dutheil
* Bug fixed when printing parameters of non-homogeneous models in bppML.

22/01/08 -*- Version 0.2.2 -*-

28/02/08 Celine Scronavacca
* Bug fixed in ReRoot.

22/01/08 -*- Version 0.2.1 -*-

06/02/08 Celine Scornavacca
* Bug fixed in ReRoot + correction of version number in bppSeqGen.

22/01/08 -*- Version 0.2.0 -*-

13/01/08 Julien Dutheil
* More options in bppml: possibility to chose between simple and double
recursion and data compression method.
* Simple recursion method improved.

11/01/08 Julien Dutheil
* Compatibility update
* bppML now output more handy parameter values.

31/12/07 Julien Dutheil
* New program Phylo Sampler.

07/11/07 Julien Dutheil
* Compatibility update with Bio++ cvs.
* Bug fixed in BppSeqGen.
* BppSeqGen can now simulate data from a set of trees (ancestral
  recombination graph).

25/10/07 Julien Dutheil
* Several bug fixed in BppML and BppSeqGen with non-homogeneous models.
* Examples updated + new example option files for non-homogeneous models.

05/10/07 Bastien Boussau & Julien Dutheil
* Non-homogenous sequence simulation in BppSeqGen.
* Model options in files have changed to suit the non-homogeneous case.
  The covarion option now is fully part of the model options, following the
	new syntax 'model=modelname+covarionname', eg 'T92+G2001'.

03/10/07 Julien Dutheil
* New program Bio++ Sequence Manipulator

21/09/07 Julien Dutheil
* Bug fixed in BppDist: branch length are indeed not estimated when parameters
  to ignore contains "BrLen".
* Stop condition when using iterations is now based on topological distance
  instead of likelihood comparisons (phyllib).
* New option optimization.topology_nni.method = [fast|better|phyml]

17/06/07 -*- Version 0.1.0 -*-

16/07/06 Julien Dutheil
* New option bootstrap.verbose for bppML.
* Bug fixed in bootstrap computation.

06/06/06 Julien Dutheil
* New program bppConsense.

05/05/07 Julien Dutheil
* First alpha release of bppsuite.

-*- Old bppML ChangeLog -*-

23/04/07 Julien Dutheil

bppSuite/bppBranchLik.cpp
// SPDX-FileCopyrightText: The Bio++ Development Group
//
// SPDX-License-Identifier: CECILL-2.1

// From the STL:
#include <iostream>
#include <iomanip>

using namespace std;

// From bpp-core:
#include <Bpp/Version.h>
#include <Bpp/Text/KeyvalTools.h>
#include <Bpp/App/ApplicationTools.h>
#include <Bpp/Text/TextTools.h>
#include <Bpp/Numeric/DataTable.h>
#include <Bpp/Numeric/VectorTools.h>

// From bpp-seq:
#include <Bpp/Seq/Alphabet/Alphabet.h>

// From bpp-phyl:
#include <Bpp/Phyl/App/BppPhylogeneticsApplication.h>
#include <Bpp/Phyl/App/PhylogeneticsApplicationTools.h>
#include <Bpp/Phyl/Tree/Tree.h>
#include <Bpp/Phyl/Tree/TreeTemplate.h>
#include <Bpp/Phyl/Likelihood/DataFlow/LikelihoodCalculationOnABranch.h>
#include <Bpp/Phyl/Likelihood/PhyloLikelihoods/OnABranchPhyloLikelihood.h>
#include <Bpp/Phyl/Likelihood/PhyloLikelihoods/PartitionProcessPhyloLikelihood.h>
#include <Bpp/Phyl/Likelihood/PhyloLikelihoods/SingleProcessPhyloLikelihood.h>
#include <Bpp/Phyl/Likelihood/PhyloLikelihoods/OneProcessSequencePhyloLikelihood.h>
#include <Bpp/Phyl/Model/AbstractBiblioMixedTransitionModel.h>
#include <Bpp/Phyl/Model/MixedTransitionModel.h>
#include <Bpp/Phyl/Model/MixtureOfATransitionModel.h>
#include <Bpp/Phyl/Model/MixtureOfTransitionModels.h>
#include <Bpp/Phyl/Model/RateDistribution/ConstantRateDistribution.h>

using namespace bpp;

/******************************************************************************/

struct rangeLik
{
  rangeLik() :
    lik_(0),
    range_()
  {}

  std::shared_ptr<LikelihoodCalculationSingleProcess> lik_;   // LikelihoodCalculationSingleProcess used
  std::vector<size_t> range_;    // vector of the positions for this process
};

struct rangeProc
{
  rangeProc() :
    proc_(0),
    range_()
  {}

  std::shared_ptr<const SubstitutionProcessInterface> proc_;   // LikelihoodCalculationSingleProcess used
  std::vector<size_t> range_;    // vector of the positions for this process
};

int main(int args, char** argv)
{
  cout << "******************************************************************" << endl;
  cout << "*     Bio++ Computation of likelihoods on branches   *" << endl;
  cout << "*                        Version " << BPP_VERSION << ".                          *" << endl;
  cout << "* Author: L. Guéguen                       Last Modif.: " << BPP_REL_DATE << " *" << endl;
  cout << "******************************************************************" << endl;
  cout << endl;

  try
  {
    BppPhylogeneticsApplication bppbranchlik(args, argv, "BppBranchLik");

    if (args == 1)
    {
      bppbranchlik.help("bppbranchlik");
      return 0;
    }

    bppbranchlik.startTimer();

    Context context;

    ///// Alphabet

    shared_ptr<const Alphabet> alphabet(bppbranchlik.getAlphabet());

    /// GeneticCode

    shared_ptr<const GeneticCode> gCode(bppbranchlik.getGeneticCode(alphabet));

    // get the data

    auto mSitesuniq = bppbranchlik.getConstAlignmentsMap(alphabet, true);

    const std::map<size_t, std::shared_ptr<const AlignmentDataInterface >> mSites = PhylogeneticsApplicationTools::uniqueToSharedMap<const TemplateAlignmentDataInterface<string>>(mSitesuniq);

    if (mSites.size() != 1)
      throw Exception("Only one alignment possible.");

    auto sites = mSites.begin()->second;
    const size_t nSites = sites->getNumberOfSites();

    /////// Get the map of initial trees

    map<string, string> unparsedParams;

    auto mpTree = bppbranchlik.getPhyloTreesMap(mSites, unparsedParams);

    /////////////////
    // Computing stuff


    shared_ptr<SubstitutionProcessCollection> SPC(bppbranchlik.getCollection(alphabet, gCode, mSites, mpTree, unparsedParams));

    auto mSeqEvoltmp = bppbranchlik.getProcesses(SPC, unparsedParams);

    auto mSeqEvol = PhylogeneticsApplicationTools::uniqueToSharedMap<SequenceEvolution>(mSeqEvoltmp);

    auto mPhyl(bppbranchlik.getPhyloLikelihoods(context, mSeqEvol, SPC, mSites));

    if (!mPhyl->hasPhyloLikelihood(0))
      throw Exception("Missing phyloLikelihoods.");

    auto tl = dynamic_pointer_cast<AlignedPhyloLikelihoodInterface>((*mPhyl)[0]);

    if (tl == 0)
      throw Exception("Only possible on aligned phyloLikelihood.");

    // Check initial likelihood:

    bppbranchlik.fixLikelihood(alphabet, gCode, tl);


    // /////////////////////////////////////////////
    // Get the alternate process

    ApplicationTools::displayMessage("\n");
    auto altprocnum = ApplicationTools::getParameter<size_t>("alt_process", bppbranchlik.getParams(), 1, "", true, false);
    ApplicationTools::displayResult("Alternative process", altprocnum);

    std::map<size_t, rangeProc> maltproc;

    if (SPC->hasSubstitutionProcessNumber(altprocnum))
    {
      rangeProc rl;
      rl.proc_ = SPC->getSubstitutionProcess(altprocnum);
      rl.range_ = std::vector<size_t>(nSites);
      std::iota(rl.range_.begin(), rl.range_.end(), 0);
      maltproc[1] = rl;
    }
    else
    {
      if (mSeqEvol.find(altprocnum) == mSeqEvol.end())
        throw BadIntegerException("Unknown alt_process number", static_cast<int>(altprocnum));

      auto seqev = mSeqEvol[altprocnum];

      auto partev = dynamic_pointer_cast<PartitionSequenceEvolution>(seqev);
      if (!partev)
        throw BadIntegerException("Unfit alt_process number: ask developers.", static_cast<int>(altprocnum));

      const auto& mapproc = partev->mapOfProcessSites();

      for (auto& proc : mapproc)
      {
        rangeProc rl;
        rl.proc_ = partev->getSubstitutionProcess(proc.first);
        rl.range_ = proc.second;
        maltproc[proc.first] = rl;
      }
    }

    // Up to now only for simple Seq Evolution


    // get the branch numbers of tree for alt process: all process must have the same numbers
    std::vector<uint> brnum;
    for (auto prn:maltproc)
    {
      const auto& brn = prn.second.proc_->parametrizablePhyloTree().getAllEdgesIndexes();
      if (brnum.size() == 0)
        brnum = brn;
      else if (brn != brnum)
        throw BadIntegerException("Non compatible trees in alternative process ", static_cast<int>(altprocnum));
    }


    // Set up segmentation of data following likelihoods
    std::map<size_t, rangeLik> mlik; // Vector of which likelihood calculation of the main phylo is assessed to each site

    auto calc = dynamic_pointer_cast<LikelihoodCalculationSingleProcess>(tl->getAlignedLikelihoodCalculation());

    if (calc)
    {
      rangeLik rl;
      rl.lik_ = calc;
      rl.range_ = std::vector<size_t>(nSites);
      std::iota(rl.range_.begin(), rl.range_.end(), 0);
      mlik[1] = rl;
    }
    else
    {
      auto parphyl = dynamic_pointer_cast<PartitionProcessPhyloLikelihood>(tl);
      if (!parphyl)
        throw Exception("Up to now, only available non Single Process is Partition. Ask developers.");

      const auto& mapproc = parphyl->getPartitionSequenceEvolution()->mapOfProcessSites();

      for (auto proc : mapproc)
      {
        rangeLik rl;
        auto pn = parphyl->getPhyloLikelihood(proc.first);
        auto opsp = dynamic_pointer_cast<OneProcessSequencePhyloLikelihood>(pn);
        auto sp = dynamic_pointer_cast<SingleProcessPhyloLikelihood>(pn);
        if (opsp)
          rl.lik_ = opsp->getLikelihoodCalculationSingleProcess();
        else if (sp)
          rl.lik_ = sp->getLikelihoodCalculationSingleProcess();
        else
          throw Exception("LikelihoodSingleProcess not available for phylolikelihood " + TextTools::toString(proc.first));
        rl.range_ = proc.second;
        mlik[proc.first] = rl;
      }
    }

    // Output Format

    string outputDesc = ApplicationTools::getStringParameter("output.lik", bppbranchlik.getParams(), "PerBranch(file=outlik.txt)");

    string outputType;
    map<string, string> outputArgs;
    KeyvalTools::parseProcedure(outputDesc, outputType, outputArgs);

    bool perSite = (outputType.find("Site") != string::npos);

    string outputFile = ApplicationTools::getStringParameter("file", outputArgs, "", "", true, 1);
    ofstream out(outputFile.c_str(), ios::out);
    out.precision(12);

    ApplicationTools::displayResult("Type of output", perSite ? "perSitePerBranch" : "PerBranch");
    ApplicationTools::displayResult("Output file", outputFile);

    // DataTable for output
    vector<string> colNames;
    shared_ptr<DataTable> rates;

    if (perSite)
    {
      colNames.push_back("Sites");

      for (auto edgeid:brnum)
      {
        colNames.push_back(TextTools::toString(edgeid));
      }

      rates = make_shared<DataTable>(nSites, colNames.size());
      rates->setColumnNames(colNames);

      // output sites
      for (unsigned int i = 0; i < nSites; i++)
      {
        const auto& currentSite = sites->site(i);
        int currentSitePosition = currentSite.getCoordinate();
        (*rates)(i, "Sites") = TextTools::toString(currentSitePosition);
      }
    }
    else
    {
      out << "Branch\tLlik\n";
    }

    // Now branch wise computation

    for (auto edgeid:brnum)
    {
      double value = 0;
      for (const auto& plik:mlik)
      {
        auto lik = plik.second.lik_;
        const auto& rangelik = plik.second.range_;

        OnABranchPhyloLikelihood obp(context, lik, edgeid);

        for (const auto& pproc:maltproc)
        {
          const auto& rangealt = pproc.second.range_;

          if ((rangealt.back() < rangelik[0]) || (rangealt[0] > rangelik.back()))
            continue;

          auto altmodel = std::unique_ptr<BranchModelInterface>(pproc.second.proc_->model(edgeid, 0).clone());
          auto confmodel = ConfiguredParametrizable::createConfigured<BranchModelInterface, ConfiguredModel>(context, std::move(altmodel));

          obp.setModel(confmodel);

          const auto& vl = obp.getLikelihoodPerSite();

          size_t poslik = 0;
          while (rangelik[poslik] < rangealt[0])
            poslik++;

          size_t posalt = 0;
          while (rangealt[posalt] < rangelik[0])
            posalt++;

          while (poslik < rangelik.size() && posalt < rangealt.size())
          {
            if (rangealt[posalt] < rangelik[poslik])
            {
              posalt++;
              continue;
            }
            if (rangelik[poslik] < rangealt[posalt])
            {
              poslik++;
              continue;
            }

            if (perSite)
              (*rates)(rangealt[posalt], TextTools::toString(edgeid)) = TextTools::toString(log(vl[poslik]), 12);
            else
              value += log(vl[poslik]);

            posalt++;
            poslik++;
          }
        }
      }

      if (!perSite)
      {
        out << edgeid << "\t" << value << "\n";
      }
    }

    if (perSite)
    {
      DataTable::write(*rates, out, "\t");
    }

    ApplicationTools::displayMessage("\n");
    bppbranchlik.done();
  }

  catch (exception& e)
  {
    cout << e.what() << endl;
    return 1;
  }

  return 0;
}
